dev/code-review/concurrency.md

```diff
@@ -0,0 +1,178 @@
+ - [Locking order for nested critical sections is documented?](
+ https://github.com/code-review-checklists/java-concurrency#document-locking-order)
+ - [Dynamically determined locks for nested critical sections are ordered?](
+ https://github.com/code-review-checklists/java-concurrency#dynamic-lock-ordering)
+ - [No extension API calls within critical sections?](
+ https://github.com/code-review-checklists/java-concurrency#non-open-call)
+
+Improving scalability
+ - [Critical section is as small as possible?](
+ https://github.com/code-review-checklists/java-concurrency#minimize-critical-sections)
+ - [Can use `ConcurrentHashMap.compute()` or Guava's `Striped` for per-key locking?](
+ https://github.com/code-review-checklists/java-concurrency#increase-locking-granularity)
+ - [Can replace blocking collection or a queue with a concurrent one?](
+ https://github.com/code-review-checklists/java-concurrency#non-blocking-collections)
+ - [Can use `ClassValue` instead of `ConcurrentHashMap<Class, ...>`?](
+ https://github.com/code-review-checklists/java-concurrency#use-class-value)
+ - [Considered `ReadWriteLock` (or `StampedLock`) instead of a simple lock?](
+ https://github.com/code-review-checklists/java-concurrency#read-write-lock)
+ - [`StampedLock` is used instead of `ReadWriteLock` when reentrancy is not needed?](
+ https://github.com/code-review-checklists/java-concurrency#use-stamped-lock)
+ - [Considered `LongAdder` instead of an `AtomicLong` for a "hot field"?](
+ https://github.com/code-review-checklists/java-concurrency#long-adder-for-hot-fields)
+
+Lazy initialization and double-checked locking
+ - [Lazy initialization of a field should be thread-safe?](
+ https://github.com/code-review-checklists/java-concurrency#lazy-init-thread-safety)
+ - [Considered double-checked locking for a lazy initialization to improve performance?](
+ https://github.com/code-review-checklists/java-concurrency#use-dcl)
+ - [Considered eager initialization instead of a lazy initialization to simplify code?](
+ https://github.com/code-review-checklists/java-concurrency#eager-init)
+ - [Double-checked locking follows the SafeLocalDCL pattern?](
+ https://github.com/code-review-checklists/java-concurrency#safe-local-dcl)
+ - [Can do lazy initialization with a benign race and without locking to improve performance?](
+ https://github.com/code-review-checklists/java-concurrency#lazy-init-benign-race)
+
+Non-blocking and partially blocking code
+ - [Non-blocking code has enough comments to make line-by-line checking as easy as possible?](
+ https://github.com/code-review-checklists/java-concurrency#check-non-blocking-code)
+ - [Can use immutable POJO + compare-and-swap operations to simplify non-blocking code?](
+ https://github.com/code-review-checklists/java-concurrency#swap-state-atomically)
+
+Threads and Executors
+ - [Thread is named?](https://github.com/code-review-checklists/java-concurrency#name-threads)
+ - [Thread is daemon?](#daemon-threads)
```

Comment:
The `#daemon-threads` and some other spots need https://github.com/code-review-checklists/java-concurrency in the link

Reply:
There are three such links, they point to Druid specific items in the bottom of the page.

Reply:
ah, got it